Multi-image overlay

kaitj opened the pull request:
Minor one here, following up on #21 - enables multi-image overlay capabilities. 

* Removes the 2 record limit prevoiusly placed on main image + overlay
* Switches overlays and overlay colormaps from single input to a list of inputs, looping over them to generate the frame as needed. If more overlays than color maps provided, it just defaults to the same color map (alternatively, we could raise an exception?)

clane9 commented on src/niclips/figures/multi_view.py:
```diff
@@ -21,21 +23,22 @@ def multi_view_frame(
     vmin: float | None = None,
     vmax: float | None = None,
-    overlay: nib.Nifti1Image | None = None,
+    overlay: list[nib.Nifti1Image] = [],
```
Might be better to support a single image or list of images, to avoid breaking compatibility?

kaitj replied:
Updated this in 47a3381 - if a single image is passed, it getes converted to a list to make use of the updated functionality addressing lists. Functionally, will still perform the same.

clane9 commented on tests/unit/niclips/figures/test_multi_view.py:
```diff
@@ -5,6 +5,7 @@
 import nibabel as nib
 import numpy as np
 import pytest
+from _pytest.logging import LogCaptureFixture
```
Is `_pytest` a typo?

kaitj replied:
I'll double check, but I think i grabbed this from another test file where the `LogCaptureFixture` was used.

kaitj commented on src/niftyone/figures/generator.py:
Ahh you're right, I missed this - will take a look.

kaitj replied:
Passed the list now instead of popping - was left over from when only a single overlay was being supported.

clane9 commented on src/niftyone/figures/generator.py:
```diff
@@ -144,17 +144,14 @@ def generate(
                 overlay = nib.nifti1.load(overlay_path)
                 overlays.append(noimg.to_iso_ras(overlay))
-                self.view_kwargs["overlay"] = overlays.pop()
+            self.view_kwargs["overlay"] = overlays
```
Note for future, and probably have said this before, but mutating `view_kwargs` on each call, which are intended to be static, is a bit of a hack.

kaitj replied:
After running into the figure name issue, I think rather than updating the view_kwargs this way, will just pass the overlay into `self.view_fn` as a kwarg. I don't think this is an issue at the moment, but to prevent any sort of issues down the road. Just need to make sure / update the `view_fn` calls to be able to accept the overlay kwarg.

kaitj replied:
Overlay is directly passed into the `view_fn` as of e16c33d rather than manipulate the `self.view_kwargs`

clane9 commented on src/niclips/figures/multi_view.py:
```diff
@@ -52,16 +59,21 @@ def multi_view_frame(
+                    "More overlays than overlay color maps- will use 'turbo'",
+                )
+            for ov, ov_cmap in zip_longest(overlay, overlay_cmap):
+                panel_overlay = noimg.render_slice(
```
The image will be annotated with XYZ coordinates by default. Not sure we need this for the overlays.

clane9 replied:
I meant how `render_slice` will add a text annotation to the image (`annotate=True` by default). In the case of overlays, this will just keep drawing e.g. `X=42, Y=10, Z=8` in the same spot.

kaitj replied:
I'll take a look - if it's already annotated once, we can have it turn off for the others.

kaitj commented on src/niclips/figures/multi_view.py:
```diff
@@ -168,10 +188,10 @@ def slice_video(
     panel_height: int | None = 256,
     cmap: str = "gray",
-    overlay_cmap: str = "brg",
+    overlay_cmap: list[str] = ["brg"],
```
If you're providing multiple overlays, you may not want to have the same cmap for them (e.g. multiple binary overlays of different tissue types). I don't think there are a lot of situations where you would want a different cmap, but the option is there if it is needed.

clane9 replied:
I just mean "brg" here vs "turbo" elsewhere. Very minor point though.

clane9 replied:
Oh, so I did this 😅 😅? Ya I can't remember any reason for doing this.

kaitj replied:
Going to leave this as is for now unless you really feel strongly about it 😛

kaitj commented on src/niclips/figures/multi_view.py:
```diff
@@ -128,7 +145,7 @@ def three_view_video(
-    overlay: nib.Nifti1Image | None = None,
+    overlay: nib.Nifti1Image | list[nib.Nifti1Image] = [],
     panel_height: int | None = 256,
     cmap: str = "gray",
```
Must have missed it here - will update.

kaitj replied:
Fixed in cc58dbb

kaitj commented on src/niclips/figures/multi_view.py:
```diff
@@ -21,25 +23,30 @@ def multi_view_frame(
     vmin: float | None = None,
     vmax: float | None = None,
-    overlay: nib.Nifti1Image | None = None,
+    overlay: nib.Nifti1Image | list[nib.Nifti1Image] = [],
```
Hmm, okay. I'll change this to keep as None instead and make the modification within the call.

kaitj commented on src/niclips/figures/multi_view.py:
Yeah I agree, though I don't anticipate overlays getting more complex than this. With the colormaps, I don't love handling it this way either. The idea of layering is sounds like a good approach to me. Would need to take a look into what this might look like.

kaitj replied:
Looked into this a little more - should be doable, though I might push this down the road and implement in a separate PR. Seems like that might be a bigger change and want to avoid this one getting too large + resolve multi-image overlays not working.

kaitj commented:
A few changes here. The main ones since last review:
  * Typecast `self.view_kwargs` to a `MappingProxyType` on `ViewGenerator` initialization. Will make it less likely to accidently mutate the kwargs once one of the generators are initialized.
  * In the same vain, passed the overlays kwarg directly to the `view_fn`.
  * Simplified generation of the figure output filepath (superceding #29). Broke down the `generate` call to a number of smaller internal calls. Also rather than the hard-to-follow entities manipulation, this has been simplified in the intiialization of the figure_entities dict.
    * Also didn't realize any non-standard entity is automatically put into `extra_entities` by b2t, so eliminated the need for setting it as a nested entity value.

Also addressed a number of the previous comments, except for:
  * Default colormaps being different.
  * Multiple overlays as layers - this is a bigger PR and will look to address this in a future PR.

kaitj commented:
Merging this one with note (see below from offline discussion) to tackle some of challenges in a subsequent PR.

From @clane9
> I think one source of issue is that the base class is trying to do a lot. all the special case logic for handling overlays and constructing file paths might be better handled in the relevant sub-classes, where the approach can also afford to be less general. For example, suppose you want a figure that overlays a tissue segmentation on a T1. You might want to incorporate some entities from the segmentation into the file name. But this is currently not possible bc the base class constructs file names based only on the first record (the T1).
